onecodex/lib/download.py

```diff
@@ -0,0 +1,80 @@
+import logging
+import os
+import os.path
+import warnings
+
+import click
+
+from onecodex.utils import FakeProgressBar
+from onecodex.exceptions import OneCodexException
+
+log = logging.getLogger("onecodex")
+
+
+def get_samples_from_project(ocx, project_name_or_id):
+    project = ocx.Projects.get(project_name_or_id)
+
+    if not project:
+        try:
+            project = ocx.Projects.where(name=project_name_or_id)[0]
+        except IndexError:
+            raise OneCodexException(
+                "No project exists with name or ID '{}'.".format(project_name_or_id)
+            )
+
+    return ocx.Samples.where(project=project)
+
+
+def filter_samples_by_tags(ocx, samples, tag_names):
+    tags = set()
+    for tag_name in tag_names:
+        log.info("Filtering samples by tag '{}'...".format(tag_name))
+        try:
+            tag = ocx.Tags.where(name=tag_name)[0]
+        except IndexError:
+            warnings.warn("No tag found: '{}'.".format(tag_name))
+            continue
+        tags.add(tag)
+
+    if tags:
+        # Filter for samples with any of the tags above (union)
+        return [sample for sample in samples if len(set(sample.tags) & tags)]
+    return samples
+
+
+def download_samples(ocx, outdir, project_name_or_id=None, tag_names=None, progressbar=False):
+    if project_name_or_id:
+        log.info("Fetching samples in project '{}'...".format(project_name_or_id))
+        samples = get_samples_from_project(ocx, project_name_or_id)
+    else:
+        log.info("Fetching all samples in your account...")
+        samples = ocx.Samples.all()
```

Comment:
Yeah it executes the API query. We let the user know what's happening in the logging message above, but you're right this might be a problem for big accounts.

@clausmith this is how the original download script worked (downloading all samples in the account if no filters provided). Should we enforce that the user provides at least one project or tag to filter by?

Reply:
No, there's a fairly common user story where a new user (e.g. a new Transnetyx customer) just wants to download all of the data from their first pilot project/sequencing run. If `len(samples) > n` though (for something like `n=50`), we should probably pop a confirmation prompt that says "Are you sure you'd like to download all `n` samples in your One Codex account?"